zapr.go

```diff
@@ -173,18 +226,69 @@ func (zl *zapLogger) GetUnderlying() *zap.Logger {
 	return zl.l
 }
 
-// newLoggerWithExtraSkip allows creation of loggers with variable levels of callstack skipping
-func newLoggerWithExtraSkip(l *zap.Logger, callerSkip int) logr.LogSink {
-	log := l.WithOptions(zap.AddCallerSkip(callerSkip))
-	return &zapLogger{
+// NewLogger creates a new logr.Logger using the given Zap Logger to log.
+func NewLogger(l *zap.Logger) logr.Logger {
+	return NewLoggerWithOptions(l)
+}
+
+// NewLogger creates a new logr.Logger using the given Zap Logger to log
+// and applies additional options.
+func NewLoggerWithOptions(l *zap.Logger, opts ...Option) logr.Logger {
+	// creates a new logger skipping one level of callstack
+	log := l.WithOptions(zap.AddCallerSkip(1))
+	zl := &zapLogger{
 		l: log,
 	}
+	zl.errorKey = "error"
+	zl.panicMessages = true
+	for _, option := range opts {
+		option(zl)
+	}
+	return logr.New(zl)
 }
 
-// NewLogger creates a new logr.Logger using the given Zap Logger to log.
-func NewLogger(l *zap.Logger) logr.Logger {
-	// creates a new logger skipping one level of callstack
-	return logr.New(newLoggerWithExtraSkip(l, 1))
+// Option is one additional parameter for NewLoggerWithOptions.
+type Option func(*zapLogger)
+
+// LogInfoLevel controls whether a numeric log level is added to
+// Info log message. The empty string disables this, a non-empty
+// string is the key for the additional field. Errors and
+// internal panic messages do not have a log level and thus
+// are always logged without this extra field.
+func LogInfoLevel(key string) Option {
+	return func(zl *zapLogger) {
+		zl.numericLevelKey = key
+	}
+}
+
+// ErrorKey replaces the default "error" field name used for the error
+// in Logger.Error calls.
+func ErrorKey(key string) Option {
+	return func(zl *zapLogger) {
+		zl.errorKey = key
+	}
+}
+
+// AllowZapFields controls whether strongly-typed Zap fields may
+// be passed instead of a key/value pair. This is disabled by
+// default because it breaks implementation agnosticism.
+func AllowZapFields(allowed bool) Option {
+	return func(zl *zapLogger) {
+		zl.allowZapFields = allowed
+	}
+}
+
+// PanicMessages controls whether extra log messages are emitted for
+// invalid log calls at zap's panic level. Depending on the
+// configuration of the zap logger, the program then panics after
+// emitting the log message which is useful in development because
+// such invalid log calls are bugs in the program. The log messages
+// explain why a call was invalid (for example, non-string
+// key). Emitting them is enabled by default.
+func PanicMessages(enabled bool) Option {
```

Comment:
Should this be "DPanicMessages" or "DPanicOnBugs"?  

a) DPanic vs Panic
b) Does it need to be more descriptive?

Consider that other libs may copy this pattern.

Reply:
I prefer `DPanicOnBugs`. It clearly ties this to the zap-specific `DPanic` and `OnBugs` makes it clear that this is only relevant for buggy log calls.

Reply:
What happens then gets explained in the comment (message and optionally panic).

Reply:
I switched to `DPanicOnBugs` for the API. Internally I kept the previous, shorter variable names and parameters.